src/c2pa.udl

```diff
@@ -2,38 +2,103 @@
 namespace c2pa {
   string version();
   string sdk_version();
+ /// [Throws=Error]
+ /// CallbackSigner create_signer(SignerCallback callback, SigningAlg alg, bytes certs, string? ta_url);
+ /// [Throws=Error]
+ /// string read_file([ByRef] string path, string? data_dir);
+ /// [Throws=Error]
+ /// string read_ingredient_file([ByRef] string path, [ByRef] string data_dir);
+ /// [Throws=Error]
+ /// sequence<u8> sign_file([ByRef] string source, [ByRef] string dest, [ByRef] string manifest, [ByRef] SignerInfo signer_info, string? data_dir);
+};
+
+[Error]
+interface Error {
+  Assertion(string reason);
+  AssertionNotFound(string reason);
+  Decoding(string reason);
+  Encoding(string reason);
+  FileNotFound(string reason);
+  Io(string reason);
+  Json(string reason);
+  Manifest(string reason);
+  ManifestNotFound(string reason);
+  NotSupported(string reason);
+  Other(string reason);
+  RemoteManifest(string reason);
+  ResourceNotFound(string reason);
+  RwLock();
+  Signature(string reason);
+  Verify(string reason);
+};
+
+enum SigningAlg {
+  "Es256",
+  "Es384",
+  "Es512",
+  "Ps256",
+  "Ps384",
+  "Ps512",
+  "Ed25519"
+};
+
+enum SeekMode {
+  "Start",
+  "End",
+  "Current"
+};
+
+callback interface Stream {
   [Throws=Error]
-  string read_file([ByRef] string path, string? data_dir);
+  bytes read_stream(u64 length);
+
   [Throws=Error]
-  string read_ingredient_file([ByRef] string path, [ByRef] string data_dir);
+  u64 seek_stream(i64 pos, SeekMode mode);
+
   [Throws=Error]
-  sequence<u8> sign_file([ByRef] string source, [ByRef] string dest, [ByRef] string manifest, [ByRef] SignerInfo signer_info, string? data_dir);
+  u64 write_stream(bytes data);
 };
 
-[Error]
-enum Error {
-  "Assertion",
-  "AssertionNotFound",
-  "Decoding",
-  "Encoding",
-  "FileNotFound",
-  "Io",
-  "Json",
-  "Manifest",
-  "ManifestNotFound",
-  "NotSupported",
-  "Other",
-  "NullParameter",
-  "RemoteManifest",
-  "ResourceNotFound",
-  "Signature",
-  "Verify"
-};
-
-dictionary SignerInfo {
-  string alg;
-  sequence<u8> sign_cert;
-  sequence<u8> private_key;
-  string? ta_url;
+interface Reader {
+  constructor();
+
+  [Throws=Error]
+  string from_stream([ByRef] string format, [ByRef] Stream reader);
+
+  [Throws=Error]
+  string json();
+
+  [Throws=Error]
+  u64 resource_to_stream([ByRef] string uri, [ByRef] Stream stream);
 };
 
+callback interface SignerCallback {
```

Comment:
It does seem like we might want a c2pa-uniffi or have this someplace where Swift and Kotlin can access it. This might be another vote for an eventual monorepo that @dyro was kicking around.

Reply:
I agree, we should find a way to share the uniffi based bindings